### ISSUE TITLE
Prefab override patch warnings in Asset Processor and Editor

### ISSUE THREAD

**Issue by forkercat** [o3de/o3de#17468]
**Describe the bug**

If AP is trying to process a prefab (e.g. SuperCar) that has an override patch that points to a deeply nested prefab (e.g. Wheel) that within a nested prefab (e.g. Car), some patch warnings will be generated.

```
Level
  SuperCar.prefab      (has an override patch that points to TireEntity)
    Car.prefab
      Wheel.prefab
        TireEntity
```

Warnings:

> The target for "replace" operation doesn't exist.
> Reason: The operation has halted during a merge operation for '/1' because a field or element has an invalid value.
> 
> Link::UpdateTarget - Some of the patches couldn't be applied on the source template 'Prefabs/Car.prefab' present under the  target Template 'Prefabs/SuperCar.prefab'.

This was first discovered in AP, but later I also notice that it could happen in Editor when you try to close the level (by opening another level or reloading the current level).

**Steps to reproduce (in Editor)**
1. Create an empty level
1. Create the testing example
  1. Create an entity `TireEntity` and create a comment component in it
  1. Create a prefab out of it called `Wheel`
  1. Create a prefab out of it called `Car`
  1. Create a prefab out of it called `SuperCar`
  1. (Focusing on SuperCar) Create a prefab override patch from SuperCar to TireEntity like changing the comment value
1. Reload the level by clicking **File - Open Recent** and select the current level

**Expected behavior**

1) No patch warnings in AP when such kind of a prefab is processed.
2) No patch warnings in Editor when a level that includes that suck kind of a prefab is closed.

**Actual behavior**

Seeing warnings in both cases.

**Screenshots/Video**

<img height="200" alt="image" src="https://github.com/o3de/o3de/assets/11157226/e5251f13-2381-4a6d-a043-d8e43eeefbec">
<img height="200" alt="image" src="https://github.com/o3de/o3de/assets/11157226/6f8bbc60-0021-472f-aed1-b77b997a7659">
<p></p>

**Found in Branch**

Development https://github.com/o3de/o3de/commit/2b0e52d2f1ca0742068cfba6bc7988141e3c57fa

**Desktop/Device:**
* Device: PC
* OS: Windows 11
* CPU: AMD Ryzen 5950x
* GPU: NVIDIA RTX 3070
* Memory: 64 GB

**Comment by forkercat** [comment]
### Cause

**tl;dr:** When updating the link between a parent prefab (e.g. SuperCar) and a nested prefab (e.g. Car), prefab system is trying to patch an entity (e.g. TireEntity) within the deeply nested prefab (e.g. Wheel) that was already deleted.

This happens in AP when PrefabBuilder is processing a prefab asset (with its nested prefabs) and after it is finished it would remove all the prefabs in memory ([code ref](https://github.com/o3de/o3de/blob/development/Gems/Prefab/PrefabBuilder/PrefabBuilderComponent.cpp#L341)).

In this case, [RemoveAllTemplates](https://github.com/o3de/o3de/blob/development/Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/PrefabSystemComponent.cpp#L709) will remove Wheel, Car, and SuperCar in this order. And when a Wheel prefab template is removed its prefab link (source: Wheel, target: Car) will also be removed, which triggers a prefab instance propagation that somehow causes the prefab system to re-apply SuperCar's override patch.

Applying the patch will fail because propagation happen later after the Wheel prefab has already been deleted. So SuperCar cannot see the TireEntity within the Wheel that is within Car.

It also happens in Editor because when a level is closed, the prefab ownership service will call `RemoveAllTemplates`.

**Comment by forkercat** [comment]
### Altering order of template removal doesn't work

Reversing and sorting the template id vector to alter the removal order doesn't not always work. A template that depends on another template doesn't always have a smaller template id. Check the below examples:

**SuperCar - Car - Wheel:**

When the PrefabLoader is loading SuperCar it gives it the id 0, 1 for Car, and 2 for Wheel.

**SuperCar - Car - Wheel / Bus - Wheel:**

In this case, the system might first load Bus then SuperCar, the id distribution would be:
- Bus - 0
- Wheel - 1
- SuperCar - 2
- Car - 3

So, we might still process the removal Wheel template before the removal of SuperCar.

### Alternative: Skip the link updates if it is a removal of all templates in PSC

Alternatively, I notice there is an existing flag called `m_removingAllTemplates` in PSC (PrefabSystemComponent) that we can use the alter the workflow. By checking this flag, we can skip some unnecessary procedures (that cause the warnings) if PSC is removing all templates.
